src/shogun/mathematics/graph/Array_test.cc

```diff
@@ -0,0 +1,49 @@
+#include <gtest/gtest.h>
+
+#include <shogun/mathematics/graph/Array.h>
+
+#include "test/GraphTest.h"
+
+#include <random>
+
+using namespace shogun;
+using namespace shogun::graph;
+using namespace std;
+
+TYPED_TEST(GraphTest, array_lvalue)
+{
+	using NumericType = typename TypeParam::c_type;
+
+	if constexpr (std::is_same_v<TypeParam, BooleanType>)
+		return;
+
+	auto X1 = SGVector<NumericType>(10);
+	auto X2 = SGVector<NumericType>(10);
+
+	X1.range_fill();
+	X2.range_fill();
+
+	auto expected_result1 = X1 + X2;
+	auto expected_result2 = expected_result1 + X2;
+
+	auto input1 = make_shared<Array>(X1);
+	auto input2 = make_shared<Array>(X2);
+	auto intermediate = input1 + input1;
+	auto output = intermediate + input2;
+
+	auto result1 =
+	    intermediate->evaluate()->template as<SGVector<NumericType>>();
+	auto result2 = output->evaluate()->template as<SGVector<NumericType>>();
```

Comment:
@karlnapf almost like JAX :) it's super inefficient right now, but it works

Reply:
if we expose this to dynamically typed interfaces this would be hidden, i.e. `numpy_array = output->evaluate()`